Dataset prerequisites

torbjorgen opened the pull request:
This PR contains prerequisites for creating a HD5 dataset for training purposes.

torbjorgen committed:
Add lib/processpool

Native multiprocessing.pool isn't stable when dealing with
long-running, high intensive CPU applications.

Adding custom written process pool for improved
handling of processes and debuggability.

Signed-off-by: Tor Björgen <tor.bjorgen@scilifelab.se>

torbjorgen committed:
Add resource usage module

This module helps in tracking subprocess resource utilization
and debugging potential OoRAM issues.

torbjorgen committed:
Add lib/hd5 viewer

Add helper method to inspect contents of a HD5 file.

Signed-off-by: Tor Björgen <tor.bjorgen@scilifelab.se>

torbjorgen committed:
ParsableVariant, VCFReader: Parse additional features in VCF file

* Allow parsing of additional modules to support building
  a complete VCF dataset.

* The custom parser fn was not properly called
  when the str contents check was relaxed.
  Make sure any custom parser is preferred
  over defaulting to generic parser.

* Provide CSQ description from VCF header

Signed-off-by: Tor Björgen <tor.bjorgen@scilifelab.se>

torbjorgen commented:
Ping @henrikstranneheim @jemten @Vince-janv 

I'd appreciate your input. Thanks!

henrikstranneheim commented on src/rdds/lib/hdf5/__main__.py:
```diff
@@ -0,0 +1,10 @@
+import argparse
+from .viewer import Hdf5Viewer
```
Best practise is to always use absolute imports and not relative

henrikstranneheim commented on src/rdds/lib/hdf5/__main__.py:
```diff
@@ -0,0 +1,10 @@
+import argparse
+from .viewer import Hdf5Viewer
+
+parser = argparse.ArgumentParser(prog='HDF5 dataset tool',
```
How about start using Click: https://click.palletsprojects.com/en/8.1.x/ it is the preferred CLI in the other Cg apps. if you agree - do the change in another PR

torbjorgen replied:
Thanks for the tip! I'll try it out next time I build a main module!

henrikstranneheim commented on src/rdds/lib/hdf5/viewer.py:
```diff
@@ -0,0 +1,45 @@
+from h5py import File, Group, Dataset
+from typing import *
```
Best practice to always be explicit with imports to not pollute the namespace i.e. avoid *.

henrikstranneheim commented on src/rdds/lib/hdf5/viewer.py:
```diff
@@ -0,0 +1,45 @@
+from typing import *
+from pprint import pprint
+from random import randint, seed
+seed(0)
```
Would be nice to explain this using a constant.

henrikstranneheim commented on src/rdds/lib/hdf5/viewer.py:
```diff
@@ -0,0 +1,45 @@
+        :return:
+        """
+
+        n_data_columns: int = 0
```
is `n` number? Then `nr` is more readable

henrikstranneheim commented on src/rdds/lib/hdf5/viewer.py:
```diff
@@ -0,0 +1,45 @@
+        n_data_columns: int = 0
+        total_ram_consumption_MB: float = 0.0
+
+        def print_data(name: str, object: object):
```
```suggestion
        def print_data(name: str, object: object) -> None:
```

henrikstranneheim replied:
Is there a more mnemonic name for object?

Is there a better description of `data`in `print_data` Data is very general.

henrikstranneheim replied:
Add doc str and include the Raise as it will potentially affect the caller of the function

henrikstranneheim commented on src/rdds/lib/hdf5/viewer.py:
```diff
@@ -0,0 +1,45 @@
+        def print_data(name: str, object: object):
+            nonlocal n_samples, h5py_file, n_data_columns, total_ram_consumption_MB
+
+            if isinstance(object, Group):
```
Maybe you should have 2 args in the function call to make object more explicit and the code cleaner.

torbjorgen replied:
The `visititems` is using the `print_data` as a callback and I don't control the argument formatting unfortunately, so I can't change this.

henrikstranneheim commented on src/rdds/lib/hdf5/viewer.py:
```diff
@@ -0,0 +1,45 @@
+                    raise NotImplementedError('Only 1D data supported at this time')
+                sample_idxs: List[int] = [randint(0, object.shape[-1]) for _ in range(0, n_samples)]
+                sample_idxs.sort()
+                data_samples = object[sample_idxs]
```
Add type hint

henrikstranneheim replied:
Could this be wrapped in a `get_data_samples` function?

henrikstranneheim commented on src/rdds/lib/hdf5/viewer.py:
```diff
@@ -0,0 +1,45 @@
+                sample_idxs: List[int] = [randint(0, object.shape[-1]) for _ in range(0, n_samples)]
+                sample_idxs.sort()
+                data_samples = object[sample_idxs]
+                dataset_size_MB: float = object.nbytes / (1024.0 * 1024.0)
```
Would be great with a constant explaining 1024. Also I wonder if you should not add the info about `_MB` in the doc str instead of in variables as you seem to only use MB.

henrikstranneheim commented on src/rdds/lib/processpool/README.md:
process_pool

henrikstranneheim commented on src/rdds/lib/processpool/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from .process_pool import ProcessPool,\
```
Do you need the `\`

torbjorgen replied:
Yes, for multi-line imports it's required.

henrikstranneheim commented on src/rdds/lib/processpool/process_pool.py:
```diff
@@ -0,0 +1,227 @@
+from threading import Thread, Lock, Event
+from multiprocessing import Process, ProcessError, cpu_count, log_to_stderr, get_context
+from multiprocessing.context import SpawnContext
+from typing import *
```
Same here

henrikstranneheim commented on src/rdds/lib/processpool/process_pool.py:
```diff
@@ -0,0 +1,227 @@
+    """
+
+    def __init__(self,
+                 fn: Callable,
```
```suggestion
                 function: Callable,
```
Explicit is better

henrikstranneheim commented on src/rdds/lib/processpool/process_pool.py:
```diff
@@ -0,0 +1,227 @@
+        """
+        self._fn = fn
+        self._args = args
+        self._n_expected_tasks: int = len(self._args)
```
```suggestion
        self._nr_expected_tasks: int = len(self._args)
```
This reads better

henrikstranneheim commented on src/rdds/lib/processpool/process_pool.py:
```diff
@@ -0,0 +1,227 @@
+        self._n_expected_tasks: int = len(self._args)
+        self._process_names = process_names
+        if self._process_names is None:
+            self._process_names: List[Any] = [None] * self._n_expected_tasks
```
```suggestion
        self._process_names = process_names if process_names else [None] * self._n_expected_tasks
```
Is this [None] a fancy way of writing 1?

henrikstranneheim commented on src/rdds/lib/processpool/process_pool.py:
```diff
@@ -0,0 +1,227 @@
+        return get_context('spawn')
+
+    @property
+    def n_expected_tasks(self) -> int:
```
```suggestion
    def nr_expected_tasks(self) -> int:
```

henrikstranneheim commented on src/rdds/lib/processpool/process_pool.py:
```diff
@@ -0,0 +1,227 @@
+        """
+        return self._n_expected_tasks
+
+    def _dispatch(self):
```
Add return hint

henrikstranneheim commented on src/rdds/lib/processpool/process_pool.py:
```diff
@@ -0,0 +1,227 @@
+            LOGGER.info(f'Dispatched task {task}')
+            self._running_tasks.append(task)
+            self._running_tasks_lock.release()
+
```
Make new function here to explain proceed and enhance readability

henrikstranneheim commented on src/rdds/lib/processpool/process_pool.py:
```diff
@@ -0,0 +1,227 @@
+        LOGGER.info('All tasks complete')
+
+    @staticmethod
+    def _threading_excepthook(except_hook_args: threading.ExceptHookArgs):
```
Add doc str with Raise

henrikstranneheim replied:
return hint

henrikstranneheim commented on src/rdds/lib/processpool/process_pool.py:
```diff
@@ -0,0 +1,227 @@
+        self._processing_complete.wait()
+
+        results: List[Task] = []
+        for i in range(0, self._n_expected_tasks):
```
```suggestion
        for _ in range(0, self._n_expected_tasks):
```

henrikstranneheim commented on src/rdds/lib/processpool/process_pool.py:
```diff
@@ -0,0 +1,227 @@
+            raise ValueError(f'Missing Task results: {results}, expected {self._n_expected_tasks} results')
+        return results
+
+    def close(self):
```
```suggestion
    def close_file_handles(self) -> None:
```

torbjorgen replied:
The native python API uses close() as the main exit routine, I prefer to keep it similar here as well.

henrikstranneheim reviewed (commented):
Did not have time to finish, but will continue later

henrikstranneheim commented on src/rdds/lib/processpool/process_pool.py:
```diff
@@ -0,0 +1,227 @@
+        LOGGER.fatal(except_hook_args)
+        raise except_hook_args.exc_value
+
+    def run_async(self) -> CompletedTaskQueue:
```
```suggestion
    def run_async_process(self) -> CompletedTaskQueue:
```

torbjorgen replied:
I'd prefer to keep this since this is closer to the native python API.

henrikstranneheim commented on src/rdds/lib/processpool/process_pool.py:
```diff
@@ -0,0 +1,227 @@
+
+        return self._completed_tasks_async_queue
+
+    def run(self) -> List[Task]:
```
```suggestion
    def run_blocking_process(self) -> List[Task]:
```

torbjorgen replied:
See comment above.

henrikstranneheim commented on src/rdds/lib/processpool/process_pool.py:
```diff
@@ -0,0 +1,227 @@
+class DummyPool:
+    """
+    Mimics the ProcessPool API but runs all processing in caller thread.
+    Good for debugging purposes.
```
```suggestion
    Used for debugging purposes.
```

henrikstranneheim commented on src/rdds/lib/resource_usage/__init__.py:
```diff
@@ -0,0 +1 @@
+from .resource_usage import ProcessResourceUsage
```
Avoid implicit imports in inint files make it explicit instead

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,92 @@
+from os import getpid
+
+
+_SCALE = {'kB': 1000.0, 'mB': 1000.0*1000.0, 'KB': 1024.0, 'MB': 1024.0*1024.0}
```
What kind of scale (put in the name)? How about using an enum  constant class instead of a dict?

torbjorgen replied:
This is used as a lookup table, so the dict is a convenient solution. Reworked the name though.

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,92 @@
+
+    def __init__(self):
+        self._pid = getpid()
+        self._proc_status = '/proc/%d/status' % self._pid
```
Use pathlib for paths and do not hard code `/`. Let pathlib handle it

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,92 @@
+        :return: Max ram consumed in Bytes
+        """
+        usage_ram_KiB: int = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
+        usage_ram_B: float = float(usage_ram_KiB * 1E3)
```
Avoid using magic numbers and use constants instead

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,92 @@
+        usage_ram_KiB: int = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
+        usage_ram_B: float = float(usage_ram_KiB * 1E3)
+        return usage_ram_B
+
```
```suggestion
        return float(usage_ram_KiB * 1E3)

```

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,92 @@
+        """
+        usage_ram_KiB: int = resource.getrusage(resource.RUSAGE_CHILDREN).ru_maxrss
+        usage_ram_B: float = float(usage_ram_KiB * 1E3)
+        return usage_ram_B
```
```suggestion
        return float(usage_ram_KiB * 1E3)
```

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,92 @@
+        usage_ram_B: float = float(usage_ram_KiB * 1E3)
+        return usage_ram_B
+
+    def _read_proc_status_virtual_memory(self, key: str):
```
Add return hint

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,92 @@
+        :return: Content in Key
+        """
+        try:
+            t = open(self._proc_status)
```
Avoid using single character names. They hard to understand and pollute the namespace.

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,92 @@
+        i = v.index(key)
+        v = v[i:].split(None, 3)  # whitespace
+        if len(v) < 3:
+            return 0.0  # invalid format?
```
Wrap in function to enable testing and encapsulation

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,92 @@
+        if len(v) < 3:
+            return 0.0  # invalid format?
+        # convert Vm value to bytes
+        return float(v[1]) * _SCALE[v[2]]
```
Make it a function and you can remove the above comment

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,92 @@
+        On a Linux machine, return current RAM (resident memory) usage.
+        :return: Current RAM usage in bytes
+        """
+        return self._read_proc_status_virtual_memory('VmRSS:')
```
Should these strings be an enum constant class for reuse and single responsibility?

torbjorgen replied:
Tried converting this into Enum class, but it fails later on in the `proc_file_contents.index(key)` with error `TypeError: must be str, not ProcFields` unfortunately.

henrikstranneheim replied:
Make it a `(str, Enum)`

torbjorgen replied:
Ah, thanks this was a great tip @henrikstranneheim ! :+1:

henrikstranneheim commented on src/rdds/lib/vcf/parsable_variant.py:
```diff
@@ -1,14 +1,45 @@
+from typing import *
 from collections import Iterator
 
+_NUMERICAL_DTYPES = [int, float, np.float32, np.float64, np.int32, np.int64]
```
What is DTYPES?

henrikstranneheim replied:
DATA_TYPES?

torbjorgen replied:
Yes, I renamed this to _NUMERICAL_DATA_TYPES.

henrikstranneheim commented on src/rdds/lib/vcf/parsable_variant.py:
```diff
@@ -1,14 +1,45 @@
 
+_NUMERICAL_DTYPES = [int, float, np.float32, np.float64, np.int32, np.int64]
+
+class ParsingError(Exception):
```
Could the name be more specific - the doc str suggests it?

torbjorgen replied:
Changed the docstring to explain that it's the base class of exceptions related to parsing.

henrikstranneheim commented on src/rdds/lib/vcf/parsable_variant.py:
```diff
@@ -20,42 +51,35 @@ def __init__(self,
-        for key, value in variant.INFO:
+        self.CHROM: str = str(variant.CHROM)  # chrom idx + x, y, mit as char
+        self.POS: int = int(variant.POS)  # integer
+        self.ID: str = str(variant.ID)  # text field
```
```suggestion
        self.CHROM = str(variant.CHROM)
        self.POS = int(variant.POS)
        self.ID: = str(variant.ID) 
```
I do not think you need to type hint when calling a constructor. This is clear enough.

henrikstranneheim commented on src/rdds/lib/vcf/parsable_variant.py:
```diff
@@ -20,42 +51,35 @@ def __init__(self,
+        # TODO: Parsing of REF/ALT fields with <ID> subtype, see https://samtools.github.io/hts-specs/VCFv4.1.pdf, p4
+        self.REF: str = str(variant.REF).replace('[', '').replace(']', '')
+        self.ALT: str = str(variant.ALT).replace('[', '').replace(']', '')
+        self._parsed_fields: List[str] = ['CHROM', 'POS', 'ID', 'REF', 'ALT']  # Placeholder of actually parsed values
```
Use constant for these VCF headers

henrikstranneheim commented on src/rdds/lib/vcf/parsable_variant.py:
```diff
@@ -20,42 +51,35 @@ def __init__(self,
+        self.REF: str = str(variant.REF).replace('[', '').replace(']', '')
+        self.ALT: str = str(variant.ALT).replace('[', '').replace(']', '')
+        self._parsed_fields: List[str] = ['CHROM', 'POS', 'ID', 'REF', 'ALT']  # Placeholder of actually parsed values
+        for key, value in list(variant.INFO):
```
```suggestion
        for field_name, field_value in list(variant.INFO):
```

henrikstranneheim commented on src/rdds/lib/vcf/parsable_variant.py:
```diff
@@ -20,42 +51,35 @@ def __init__(self,
-                    # Set private attribute as fallback, to allow runtime inspection
-                    self.__setattr__('_unparsed_'+key, value)
+            try:  # custom parser
+                parsing_fn: callable = self.__getattribute__('_parse_fn_%s' % key)
```
Not a fan of accessing private variables like this. Can one use vars() or something similar?

henrikstranneheim commented on src/rdds/lib/vcf/parsable_variant.py:
```diff
@@ -20,42 +51,35 @@ def __init__(self,
-                    self.__setattr__('_unparsed_'+key, value)
+            try:  # custom parser
+                parsing_fn: callable = self.__getattribute__('_parse_fn_%s' % key)
+                if parsing_fn is not None:
```
```suggestion
                if parsing_fn:
```
Would this work or does it need to this specific?

henrikstranneheim commented on src/rdds/lib/vcf/parsable_variant.py:
```diff
@@ -20,42 +51,35 @@ def __init__(self,
+            except AttributeError:
+                pass
+
+            self._parse_store_data(key, value)
```
More than 1 arg - use named args

henrikstranneheim commented on src/rdds/lib/vcf/parsable_variant.py:
```diff
@@ -69,64 +93,180 @@ def __iter__(self) -> Iterator:
+        """
+        Return parsed data type for a VCF field.
+        :param field_name: VCF field name (parsed)
+        :return: Type
```
Add raise to doc str as it affects the one calling the function

henrikstranneheim commented on src/rdds/lib/vcf/parsable_variant.py:
```diff
@@ -69,64 +93,180 @@ def __iter__(self) -> Iterator:
+        except AttributeError:
+            raise DtypeNotFoundError(f'Field name \'{field_name}\' not found')
+
+    def _parse_store_data(self,
```
Could this functon call smaller functions to make it cleaner it is not so easy to read?

henrikstranneheim commented on src/rdds/lib/vcf/parsable_variant.py:
```diff
@@ -69,64 +93,180 @@ def __iter__(self) -> Iterator:
-        self.__setattr__('RankScore_family_id', family_id)
-        self.__setattr__('RankScore_value', rank_score_value)
-        self._parsed_fields.extend(['RankScore_family_id', 'RankScore_value'])
+        self._parse_store_data(key='RankScore_family_id', data=parts[0], expected_dtype=str)
```
Use constant to make sure consistent usage and single responsability

henrikstranneheim commented on src/tests/processpool/test_processpool.py:
```diff
@@ -0,0 +1,182 @@
+    Test pool restart capability.
+    """
+    test_processpool()
+    test_processpool()
```
?

torbjorgen replied:
This is a restart pool test, making sure no resources are still allocated and blocking the new pool from initializing.

henrikstranneheim reviewed (commented):
Complex stuff 👍 . To big PR  to do in one review. Will comment now and take another round when you feel you have addressed the comment. It would be great for future PRs if the scope can be smaller. It will make reviewing a lot easier 😄

torbjorgen commented:
> Complex stuff 👍 . To big PR to do in one review. Will comment now and take another round when you feel you have addressed the comment. It would be great for future PRs if the scope can be smaller. It will make reviewing a lot easier 😄

Thanks for great feedback @henrikstranneheim !

I've reworked the PR now based on your feedback.

Sure thing, I'll try to reduce the PR size even further going forward.
I see now that this PR could have been split on modules for example.

henrikstranneheim commented on src/rdds/lib/hdf5/viewer.py:
```diff
@@ -0,0 +1,61 @@
+from h5py import File, Group, Dataset
+from typing import List, Union
+import numpy as np
```
```suggestion
import numpy
```
Keep the namespace clean and explicit 😄

torbjorgen replied:
I'm trying to stick to the numpy convention described here: https://numpy.org/doc/stable/user/absolute_beginners.html#how-to-import-numpy

henrikstranneheim commented on src/rdds/lib/hdf5/viewer.py:
```diff
@@ -0,0 +1,61 @@
+from random import randint, seed
+seed(0)  # Make sure to sample data from dataset identically every time.
+
+KIBIBYTE: float = 1024.0
```
```suggestion
KIBIBYTE: float = 1024.0
```
What is a KIBIBYTE?

torbjorgen replied:
It's the binary (in power of 2) equivalent to KILOBYTE (1E3 bytes).

henrikstranneheim commented on src/rdds/lib/hdf5/viewer.py:
```diff
@@ -0,0 +1,61 @@
+
+
+class Hdf5Viewer:
+
```
Add class doc str

henrikstranneheim commented on src/rdds/lib/hdf5/viewer.py:
```diff
@@ -0,0 +1,61 @@
+            """
+            Prints subset of data in Dataset, or Group name.
+
+            Only support printing 1D data, raises NotImplemtedError.
```
```suggestion
            Only support printing 1D data.
```

henrikstranneheim commented on src/rdds/lib/hdf5/viewer.py:
```diff
@@ -0,0 +1,61 @@
+                sample_idxs: List[int] = [randint(0, object.shape[-1]) for _ in range(0, n_samples)]
+                sample_idxs.sort()
+                data_samples: np.ndarray = object[sample_idxs]
+                return data_samples
```
```suggestion
                return object[sample_idxs]
```

henrikstranneheim commented on src/rdds/lib/hdf5/viewer.py:
```diff
@@ -0,0 +1,61 @@
+                nr_data_columns += 1
+                if len(object.shape) > 1:
+                    raise NotImplementedError('Only 1D data supported at this time')
+                data_samples = get_samples()
```
```suggestion
                samples = get_samples()
```
Add type hint

henrikstranneheim commented on src/rdds/lib/process_pool/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from rdds.lib.process_pool.process_pool import ProcessPool,\
```
I think it is good to avoid these implicit import - they mask where code is being imported from and therefore pollute the namespace. Import where used

torbjorgen replied:
The design idea behind these imports is that the `__init__.py` defines the public API of the module.

I think this makes module refactoring easier going forward;  users of the module otherwise has to rely on the full internal path? @henrikstranneheim

henrikstranneheim replied:
Exactly - explicit and no chance for confusion and namemangling. Also your IDE should handle imports suggestions automatically so one rarely writes an import path

henrikstranneheim commented on src/rdds/lib/process_pool/process_pool.py:
```diff
@@ -0,0 +1,235 @@
+
+        def dispatch_new_process() -> None:
+            self._running_tasks_lock.acquire()
+            process: Process = self._ctx.Process(target=self._function,
```
```suggestion
            process = self._ctx.Process(target=self._function,
```
No need to type hint when calling a constructor. It is already clear what the return will be.

henrikstranneheim commented on src/rdds/lib/process_pool/process_pool.py:
```diff
@@ -0,0 +1,235 @@
+                                                 args=process_args,
+                                                 name=process_name)
+            process.start()
+            task: Task = Task(args=process_args, process=process)
```
```suggestion
            task = Task(args=process_args, process=process)
```

henrikstranneheim commented on src/rdds/lib/process_pool/process_pool.py:
```diff
@@ -0,0 +1,235 @@
+        self._dispatch_thread = Thread(target=self._dispatch,
+                                       name='dispatch')
+        self._collect_stopped_tasks_thread = Thread(target=self._collect_stopped_tasks,
+                                                    name='collect_stopped_processes')
```
It would be good to use a constant to name: 'collect_stopped_processes' and thereby describe it a bit more

torbjorgen replied:
Sorry, I do not understand your suggestion @henrikstranneheim . Can you elaborate a bit more? 
Is the `_thread` suffix not descriptive enough?

henrikstranneheim replied:
I was looking to not hard code 'collect_stopped_processes'

torbjorgen replied:
Opened a fix in #15 @henrikstranneheim

henrikstranneheim commented on src/rdds/lib/process_pool/process_pool.py:
```diff
@@ -0,0 +1,235 @@
+
+        results: List[Task] = []
+        for _ in range(0, self._nr_expected_tasks):
+            results.append(async_results.get(timeout=1))  # Throws EmptyException
```
```suggestion
            results.append(async_results.get(timeout=1))
```

henrikstranneheim commented on src/rdds/lib/process_pool/process_pool.py:
```diff
@@ -0,0 +1,235 @@
+        self._args = args
+        self._nr_expected_tasks = len(self._args)
+        self._process_names = process_names
+        if self._process_names is None:
```
```suggestion
        if not self._process_names:
```

henrikstranneheim commented on src/rdds/lib/process_pool/process_pool.py:
```diff
@@ -0,0 +1,235 @@
+            LOGGER.info(f'[DummyPool] Dispatching task {process_name}')
+            self._function(*arg)
+            # If function returned to callee, then it's successfully completed.
+            task: Task = Task(args=arg, process=DummyProcess(name=process_name, exitcode=0))
```
```suggestion
            task = Task(args=arg, process=DummyProcess(name=process_name, exitcode=0))
```

henrikstranneheim commented on src/rdds/lib/resource_usage/__init__.py:
```diff
@@ -0,0 +1 @@
+from rdds.lib.resource_usage.resource_usage import ProcessResourceUsage
```
Avoid implicit imports

torbjorgen replied:
Please see my comment on this in process_pool.

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,108 @@
+        :param key: Key to fetch
+        :return: Parsed value, as float
+        """
+        i = proc_file_content.index(key)
```
What does i represent? Can we use that as a name?

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,108 @@
+        i = proc_file_content.index(key)
+        entry: List[str] = proc_file_content[i:].split(None, 3)  # whitespace removal
+        if len(entry) < 3:
+            return 0.0  # invalid format?
```
```suggestion
            return 0.0  # invalid format
```
Are you asking or telling the reader about the format?

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,108 @@
+            proc_file_content: str = proc_file.read()
+            proc_file.close()
+        except:
+            return 0.0  # non-Linux?
```
Same here - maybe make an invalid constant and skip the comment.

henrikstranneheim commented on src/rdds/lib/resource_usage/resource_usage.py:
```diff
@@ -0,0 +1,108 @@
+        s += f'RSS:{(self.get_current_rss() / GIGABYTE):.2f}gB({(self.get_max_rss() / GIGABYTE):.2f}gB), '
+        s += f'VM:{(self.get_current_vm() / GIGABYTE):.2f}gB({(self.get_max_vm() / GIGABYTE):.2f}gB), '
+        s += f'Stack:{(self.get_stack_size_bytes() / MEGABYTE):.2f}mB, '
+        return s
```
What does s stand for?

henrikstranneheim commented on src/rdds/lib/vcf/vcf.py:
```diff
@@ -1,6 +1,6 @@
 import cyvcf2
 from cyvcf2 import Variant
-
+from typing import *
```
Avoid using `*` - it will pollute the namespace

henrikstranneheim commented on src/rdds/lib/vcf/vcf.py:
```diff
@@ -47,3 +47,73 @@ def number_of_variants(self) -> int:
+        Static fields that's always expected to be present in VCF
+        :return:
+        """
+        return ['CHROM', 'POS', 'ID', 'REF', 'ALT']
```
Define these as a string bases Enum class instead for reuse

henrikstranneheim commented on src/rdds/lib/vcf/vcf.py:
```diff
@@ -47,3 +47,73 @@ def number_of_variants(self) -> int:
+        Return all FORMAT fields
+        :return:
+        """
+        return self._get_type_fields(type='FORMAT')
```
Belong in Enum ckass

henrikstranneheim commented on src/rdds/lib/vcf/vcf.py:
```diff
@@ -47,3 +47,73 @@ def number_of_variants(self) -> int:
+        Return all INFO fields
+        :return:
+        """
+        return self._get_type_fields(type='INFO')
```
Same here

henrikstranneheim commented on src/rdds/lib/vcf/vcf.py:
```diff
@@ -47,3 +47,73 @@ def number_of_variants(self) -> int:
+        return data_fields
+
+    @property
+    def csq_description(self) -> Union[str, None]:
```
```suggestion
    def csq_description(self) -> Optional[str]:
```

henrikstranneheim commented on src/rdds/lib/vcf/vcf.py:
```diff
@@ -47,3 +47,73 @@ def number_of_variants(self) -> int:
+                pass
+
+    @property
+    def csq_sub_fields(self) -> Union[List[str], None]:
```
```suggestion
    def csq_sub_fields(self) -> Optional[List[str]]:
```

henrikstranneheim commented on src/tests/process_pool/test_processpool.py:
```diff
@@ -0,0 +1,182 @@
+import numpy as np
+import pytest as pt
```
```suggestion
import numpy
import pytest
```

torbjorgen replied:
I'm trying to stick to the numpy convention described here: https://numpy.org/doc/stable/user/absolute_beginners.html#how-to-import-numpy

henrikstranneheim commented on src/tests/process_pool/test_processpool.py:
```diff
@@ -0,0 +1,182 @@
+
+
+# Processing functions must be accessible from separate process (i.e. not inside local scope in test method)
+def f_return_value(x: int, q: SimpleQueue):
```
Avoid using single character variable names

henrikstranneheim commented on src/tests/process_pool/test_processpool.py:
```diff
@@ -0,0 +1,182 @@
+
+
+def f_erroneous(*args, **kwargs):
+    raise RuntimeError('Whoops')
```
Something more descriptive would be nice 😄

henrikstranneheim commented on src/tests/process_pool/test_processpool.py:
```diff
@@ -0,0 +1,182 @@
+    # THEN expect all items to be processed
+    for i in range(0, len(args)):
+        result_sum += q.get()
+    for i in range(0, len(args)):
```
```suggestion
    for _ in range(0, len(args)):
        result_sum += q.get()
    for _ in range(0, len(args)):
```

henrikstranneheim commented on src/tests/resource_usage/test_resource_usage.py:
```diff
@@ -0,0 +1,16 @@
+    stats = ProcessResourceUsage()
+    # WHEN reading the runtime statistics
+    # THEN expect it to succeed
+    _ = '%s' % str(stats)
```
assert?

henrikstranneheim commented on src/tests/vcf/test_parsable_variant.py:
```diff
@@ -0,0 +1,20 @@
+    # GIVEN a VCF file
+    # WHEN reading it
+    vcf_reader: VCFReader = VCFReader(EXAMPLE_FILE)
+    assert vcf_reader.number_of_variants == 390  # 500 rows, 110 is header, remaining is the variant data
```
Use constant instead of header

henrikstranneheim commented on src/tests/vcf/test_parsable_variant.py:
```diff
@@ -0,0 +1,20 @@
+    assert vcf_reader.number_of_variants == 390  # 500 rows, 110 is header, remaining is the variant data
+    csq_description: str = vcf_reader.csq_description
+    for n_variants, unparsed_variant in enumerate(vcf_reader):
+        _: ParsableVariant = ParsableVariant(variant=unparsed_variant,
```
```suggestion
        _ = ParsableVariant(variant=unparsed_variant,
```

henrikstranneheim commented on src/tests/vcf/test_vcf.py:
```diff
@@ -1,13 +1,18 @@
 from os.path import dirname
-EXAMPLE_FILE: str = dirname(__file__)+'/test.vcf'
+EXAMPLE_FILE: str = dirname(__file__)+'/test.vcf.gz'
```
I think these can be added as test fixture files and then you can avoid using the private `__file__`. See cg for an example: `tests/fixtures`

torbjorgen replied:
Good point. I'll probably have to revise this schema once the amount of test data increases.

I'll rework this in a separate PR.

henrikstranneheim reviewed (approved):
Massive 👍

torbjorgen commented:
* Updated according to @henrikstranneheim suggestions.

torbjorgen commented:
* Created new issue related to pytest file fixtures
https://github.com/Clinical-Genomics/rdds/issues/14
